pkg/cloud/services/network/natgateways_test.go

```diff
@@ -40,6 +40,161 @@ const (
 	ElasticIPAllocationID = "elastic-ip-allocation-id"
 )
 
+func TestFindNatGatewayForEdgeSubnet(t *testing.T) {
+	knownSubnets := []infrav1.SubnetSpec{
+		{
+			ID:               "subnet-1",
+			AvailabilityZone: "us-east-1a",
+			CidrBlock:        "10.0.10.0/24",
+			IsPublic:         false,
+			NatGatewayID:     aws.String("nat-1234567890abcdef0"),
+		},
+		{
+			ID:               "subnet-2",
+			AvailabilityZone: "us-east-1b",
+			CidrBlock:        "10.0.20.0/24",
+			IsPublic:         false,
+		},
+		{
+			ID:               "subnet-3",
+			AvailabilityZone: "us-east-1c",
+			CidrBlock:        "10.0.30.0/24",
+			IsPublic:         true,
+			NatGatewayID:     aws.String("nat-1234567890abcdef1"),
+		},
+		{
+			ID:               "subnet-4",
+			AvailabilityZone: "us-east-1d",
+			CidrBlock:        "10.0.40.0/24",
+			IsPublic:         true,
+			NatGatewayID:     aws.String("nat-1234567890abcdef2"),
+			ParentZoneName:   aws.String("zone-subnet-4"),
+		},
+	}
+
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	testCases := []struct {
+		name         string
+		input        infrav1.SubnetSpec
+		expectErr    bool
+		expect       string
+		knownSubnets []infrav1.SubnetSpec
+	}{
+		{
+			name: "return an error if there are no nat gateways available",
+			input: infrav1.SubnetSpec{
+				AvailabilityZone: "us-east-1a",
+				CidrBlock:        "10.0.10.0/24",
+			},
+			expectErr: true,
+		},
+		{
+			name: "return an error if the input subnet is public",
+			input: infrav1.SubnetSpec{
+				CidrBlock:  "10.0.10.0/24",
+				ResourceID: "subnet-1234567890abcdef2",
+				IsPublic:   true,
+			},
+			expectErr: true,
+		},
+		{
+			name: "return the ID of the first public subnet's NAT gateway associated with the input subnet",
+			input: infrav1.SubnetSpec{
+				AvailabilityZone: "us-east-1a",
+				CidrBlock:        "10.0.10.0/24",
+			},
+			expect: *knownSubnets[2].NatGatewayID,
+		},
+		{
+			name: "return nat gateway ID for the input subnet with a parent zone name",
+			input: infrav1.SubnetSpec{
+				CidrBlock:      "10.0.10.0/24",
+				ParentZoneName: aws.String("us-east-1d"),
+			},
+			expect: *knownSubnets[3].NatGatewayID,
+		},
+		{
+			name: "return nat gateway ID for the input subnet with a parent zone name and availability zone",
+			input: infrav1.SubnetSpec{
+				CidrBlock:        "10.0.10.0/24",
+				AvailabilityZone: "us-east-1c",
```

Comment:
I would change the zones with parent to Local Zone name. Example:
```suggestion
				AvailabilityZone: "us-east-1-bos-1a",
```